_posts/2021-04-28-#17331.md

```diff
@@ -0,0 +1,111 @@
+---
+layout: pr
+date: 2021-04-28
+title: "Use effective values throughout coin selection"
+pr: 17331
+authors: [achow101]
+components: ["wallet"]
+host: Xekyo
+status: upcoming
+commit: 4ac1add
+---
+
+## Notes
+
+- Wallets maintain their balance as a pool of Unspent Transaction Outputs
+  (UTXOs).
+
+- _Coin selection_ refers to the process of picking UTXOs from the wallet's UTXO
+  pool to fund a transaction.
+
+- Beyond the primary goal of funding a transaction, the secondary goals of coin
+  selection are:
+
+    - Minimizing short term and long term fees:  
+      We want to both pay the least right now, but we also want to be good
+      stewards of our UTXO pool and not rack up large UTXO counts that will
+      translate to huge future transaction fees.
+
+    - Maintaining financial privacy:  
+      Ideally, our transactions leak as little information about ourselves and
+      our business partners as possible.
+
+    - Reliably move the payment(s) to finalization:  
+      Reduce the chance of transactions being conflicted and transfer value in a
+      limited time-frame
+
+- Using unconfirmed inputs can make transactions unreliable. When a foreign
+  input is used, the parent transaction may time out or get replaced. Even using
+  self-sent unconfirmed funds may delay the new transaction if the parent
+  transaction has an extensive ancestry, is extraordinarily large, or used a
+  lower feerate than targeted for the child transaction.
+
+- When a new transaction is built, the target feerate is provided as an input.
```

Comment:
I think a more helpful note would be a rough outline of the process of creating a tx, with a bit of extra detail for coin selection specifically:

1. `CreateTransaction()` builds the header and payment outputs (no change yet), which are fixed.
2. Grab feerates using `GetMinimumFeeRate()` (goes through fee estimation and/or user configurations). All of these are fixed: the tx header, payment outputs, target feerate, target amount, etc.
3. Get all `AvailableCoins()`
4. Use `SelectCoins()` to get the inputs (fund the transaction). 
  (4a) This uses pre-selected coins first (`CoinControl`), then groups UTXOs into `OutputGroup`s, and runs `SelectCoinsMinConf()` with increasingly permissive `CoinEligibilityFilter`s until we find a set of coins.
   (4b) Each iteration of `SelectCoinsMinConf()` tries BnB, then Knapsack if that fails.
5. Sign the transaction

Reply:
Thanks, this was super helpful. I've added it with some slight modifications.